File: legacy/_1.7.10/src/main/java/cn/enaium/noexpensive/callback/EnchantmentCanCombineCallback.java
package cn.enaium.noexpensive.callback;

import net.legacyfabric.fabric.api.event.Event;
import net.legacyfabric.fabric.api.event.EventFactory;
import net.minecraft.enchantment.Enchantment;

/**
 * @author Enaium
 */
public interface EnchantmentCanCombineCallback {
    Event<EnchantmentCanCombineCallback> EVENT = EventFactory.createArrayBacked(EnchantmentCanCombineCallback.class, (listeners) -> (enchantment1, enchantment2) -> {
        for (EnchantmentCanCombineCallback listener : listeners) {
            boolean result = listener.interact(enchantment1, enchantment2);
            if (!result) {
                return false;
            }
        }

        return true;
    });

    boolean interact(Enchantment enchantment1, Enchantment enchantment2);
}

File: legacy/_1.8.9/src/main/kotlin/cn/enaium/noexpensive/callback/EnchantmentCanCombineCallback.kt
package cn.enaium.noexpensive.callback

import net.legacyfabric.fabric.api.event.EventFactory
import net.minecraft.enchantment.Enchantment

/**
 * @author Enaium
 */
fun interface EnchantmentCanCombineCallback {
    fun interact(enchantment1: Enchantment, enchantment2: Enchantment): Boolean

    companion object {
        val EVENT = EventFactory.createArrayBacked(
            EnchantmentCanCombineCallback::class.java
        ) { listeners: Array<EnchantmentCanCombineCallback> ->
            EnchantmentCanCombineCallback { enchantment1: Enchantment, enchantment2: Enchantment ->
                for (listener in listeners) {
                    val result = listener.interact(enchantment1, enchantment2)
                    if (!result) {
                        return@EnchantmentCanCombineCallback false
                    }
                }
                return@EnchantmentCanCombineCallback true
            }
        }
    }
}
